commons/com.b2international.index/src/com/b2international/index/es/client/EsClient.java

```diff
@@ -0,0 +1,148 @@
+import org.elasticsearch.index.reindex.BulkByScrollResponse;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.transport.client.PreBuiltTransportClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.b2international.index.Activator;
+import com.b2international.index.es.EsClientConfiguration;
+import com.b2international.index.es.client.http.EsHttpClient;
+import com.b2international.index.es.client.tcp.EsTcpClient;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.cache.RemovalNotification;
+import com.google.common.net.HostAndPort;
+import com.google.common.util.concurrent.UncheckedExecutionException;
+
+/**
+ * @since 6.11
+ */
+public interface EsClient extends AutoCloseable {
+
+	Logger LOG = LoggerFactory.getLogger("elastic-snowowl");
+	
+	IndicesClient indices();
+	
+	ClusterClient cluster();
+	
+	GetResponse get(GetRequest req) throws IOException;
+	
+	SearchResponse search(SearchRequest req) throws IOException;
+	
+	SearchResponse scroll(SearchScrollRequest req) throws IOException;
+	
+	ClearScrollResponse clearScroll(ClearScrollRequest req) throws IOException;
+	
+	BulkProcessor.Builder bulk(BulkProcessor.Listener listener);
+	
+	BulkByScrollResponse updateByQuery(String index, String type, int batchSize, Script script, int numberOfSlices, QueryBuilder query) throws IOException;
+	
+	static EsClient create(final EsClientConfiguration configuration) {
+		return ClientPool.create(configuration);
+	}
+	
+	static void closeAll() {
+		ClientPool.closeAll();
+	}
+	
+	/**
+	 * @since 6.11
+	 */
+	final class ClientPool {
+		
+		private static final LoadingCache<EsClientConfiguration, EsClient> CLIENTS_BY_HOST = CacheBuilder.newBuilder()
+				.removalListener(ClientPool::onRemove)
+				.build(CacheLoader.from(ClientPool::onAdd));
+		
+		private ClientPool() {}
+		
+		static EsClient create(EsClientConfiguration configuration) {
+			try {
+				return CLIENTS_BY_HOST.getUnchecked(configuration);
+			} catch (UncheckedExecutionException e) {
+				if (e.getCause() instanceof RuntimeException) {
+					throw (RuntimeException) e.getCause();
+				} else {
+					throw new RuntimeException(e.getCause());
+				}
+			}
+		}
+
+		static void closeAll() {
+			CLIENTS_BY_HOST.invalidateAll();
+			CLIENTS_BY_HOST.cleanUp();
+		}
+		
+		@SuppressWarnings("resource")
+		static EsClient onAdd(final EsClientConfiguration configuration) {
+			LOG.info("Connecting to Elasticsearch cluster at '{}'{}, connect timeout: {} ms, socket timeout: {} ms.", 
+					configuration.getClusterUrl(),
+					configuration.isProtected() ? " using basic authentication" : "",
+					configuration.getConnectTimeout(),
+					configuration.getSocketTimeout());
+			
+			if (configuration.isHttp()) {
+				return new EsHttpClient(configuration);
+			} else {
+				checkState(configuration.isTcp(), "Only TCP and HTTP clients are supported");
```

Comment:
We could also throw an exception if the transport client is configured with a user name and password, as that is only supported in the x-pack version.

Reply:
See [Java Client and Security](https://www.elastic.co/guide/en/elastic-stack-overview/6.5/java-clients.html) in the Elastic Stack documentation.